src/functions.php

```diff
@@ -505,8 +506,9 @@ function extract_read_preference_from_options(array $options): ?ReadPreference
 function select_server(Manager $manager, array $options): Server
 {
     $session = extract_session_from_options($options);
-    if ($session instanceof Session && $session->getServer() !== null) {
-        return $session->getServer();
+    $server = $session instanceof Session ? $session->getServer() : null;
+    if ($server !== null) {
+        return $server;
```

Comment:
Interestingly enough, psalm was quite right to complain about this. Even if the first call to `getServer()` in the conditional returns a `Server` instance, it's not guaranteed that the second call in the `return` statement will return the same `Server` object, as the session could've been unpinned in between evaluating the conditional and executing the `return`. In that case, `getServer()` would return `null` which would be an invalid return value triggering a fatal error. (I'll leave the exercise of proving me wrong to the curious reader).

Reply:
Since `mongoc_client_session_get_server_id()` returns the `server_id` property of the `mongoc_client_session_t` struct, I'm not sure how you'd be able to interrupt this without some hacking approach like runtime debugging or `register_tick_function()`.

That said, avoiding a second call to `getServer()` seems entirely reasonable.